msal/public_client_application.go

```diff
@@ -14,24 +18,32 @@ type PublicClientApplication struct {
 	clientApplication *clientApplication
 }
 
-// CreatePublicClientApplication creates a PublicClientApplication instance given a client ID and authority URL.
-func CreatePublicClientApplication(clientID string, authority string) (*PublicClientApplication, error) {
-	clientApp := createClientApplication(clientID, authority)
-	pca := &PublicClientApplication{
-		clientApplication: clientApp,
-	}
-	return pca, nil
+// PublicClientApplicationOptions configures the PublicClientApplication's behavior.
+type PublicClientApplicationOptions struct {
+	// Accessor controls cache persistence.
+	// By default there is no cache persistence.
+	Accessor CacheAccessor
+
+	// Client sets the transport for making HTTP requests.
+	// Leave this as nil to use the default HTTP transport.
+	Client HTTPClient
```

Comment:
Could we just use an http.Client? and users could substitute either a custom httpClient or just the Transport?

Reply:
We could, but it tightly couples us to this specific implementation.

In the track 2 Go SDK, we use a pipeline middleware that lets us (and SDK consumers) have more control over how HTTP requests/responses are handled.  If we keep this as a generic interface it would allow consumers to inject their own middleware (ideally I'd like to see us being able to use pipeline policies we have in `azcore` as we provide mechanisms like retries, logging, telemetry etc; see [here](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore) for more info).

I wanted to start simple for now, and refine it later.

Reply:
not sure if this is the best way to pass in the HttpClient, but certainly a functionality we should support like we do for other libraries

---
In reply to: [514778150](https://github.com/AzureAD/microsoft-authentication-library-for-go/pull/68#discussion_r514778150) [](ancestors = 514778150)

Reply:
Aside: why is the object not called HttpClient, client can be so many things

---
In reply to: [515451140](https://github.com/AzureAD/microsoft-authentication-library-for-go/pull/68#discussion_r515451140) [](ancestors = 515451140,514778150)

Reply:
Renamed to `HTTPClient`